src/catalog/catalog.cpp

```diff
@@ -930,6 +935,287 @@ std::shared_ptr<SystemCatalogs> Catalog::GetSystemCatalogs(
+ * @param txn the transaction Context
+ * @return TransactionContext ResultType(SUCCESS or FAILURE)
+ */
+ResultType Catalog::AlterTable(oid_t database_oid, oid_t table_oid, const std::string &schema_name,
+                               std::unique_ptr<catalog::Schema> &new_schema,
+                               concurrency::TransactionContext *txn) {
+  LOG_TRACE("AlterTable in Catalog");
+
+  if (txn == nullptr)
+    throw CatalogException("Alter table requires transaction");
+  try {
+    auto storage_manager = storage::StorageManager::GetInstance();
+    auto database = storage_manager->GetDatabaseWithOid(database_oid);
+    try {
+      auto old_table = database->GetTableWithOid(table_oid);
+      auto old_schema = old_table->GetSchema();
+      auto pg_index = catalog_map_[database_oid]->GetIndexCatalog();
+
+      // Step 1: build empty table with new schema
+      bool own_schema = true;
+      bool adapt_table = false;
+      auto new_table = storage::TableFactory::GetDataTable(
+          database_oid, table_oid,
+          catalog::Schema::CopySchema(new_schema.get()), old_table->GetName(),
+          DEFAULT_TUPLES_PER_TILEGROUP, own_schema, adapt_table);
+      // Step 2: Copy indexes
+      auto old_index_oids = pg_index->GetIndexObjects(table_oid, txn);
+      for (auto index_oid_pair : old_index_oids) {
+        oid_t index_oid = index_oid_pair.first;
+        // delete record in pg_index
+        pg_index->DeleteIndex(index_oid, txn);
+        // Check if all indexed columns still exists
+        auto old_index = old_table->GetIndexWithOid(index_oid);
+        bool index_exist = true;
+        std::vector<oid_t> new_key_attrs;
+
+        for (oid_t column_id : old_index->GetMetadata()->GetKeyAttrs()) {
+          bool is_found = false;
+          std::string column_name = old_schema->GetColumn(column_id).GetName();
+          oid_t i = 0;
+          for (auto new_column : new_schema->GetColumns()) {
+            if (column_name == new_column.GetName()) {
+              is_found = true;
+              new_key_attrs.push_back(i);
+              break;
+            }
+            i++;
+          }
+          if (!is_found) {
+            index_exist = false;
+            break;
+          }
+        }
+        if (!index_exist) continue;
+
+        // construct index on new table
+        auto index_metadata = new index::IndexMetadata(
+            old_index->GetName(), index_oid, table_oid, database_oid,
+            old_index->GetMetadata()->GetIndexType(),
+            old_index->GetMetadata()->GetIndexConstraintType(),
+            new_schema.get(),
+            catalog::Schema::CopySchema(new_schema.get(), new_key_attrs),
+            new_key_attrs, old_index->GetMetadata()->HasUniqueKeys());
+
+        std::shared_ptr<index::Index> new_index(
+            index::IndexFactory::GetIndex(index_metadata));
+        new_table->AddIndex(new_index);
+
+        // reinsert record into pg_index
+        pg_index->InsertIndex(
```

Comment:
[race condition]
It is risky to manipulate index directly, because it is not protected by any consistency control mechanism like locks. It is better to do such operations through an executor just like your step 4(use a SeqScanPlan). The index executor with lock mechanism will take care of consistency.

Reply:
Good idea!